### ISSUE TITLE
`Alert` `send_keys` method doesn't work on iOS

### ISSUE THREAD

**Issue by SalmonMode** [appium/appium#7585]
## The problem

If I have an alert present (`XCUIElementTypeAlert`), I can grab the alert, and even access its `.text` property, but `alert.send_keys("something")` throws the following error:

```
(Pdb) alert = self.driver.switch_to_alert()
(Pdb) alert.send_keys("something")
*** WebDriverException: Message: An unknown server-side error occurred while processing the command. Original error: Unhandled endpoint: /session/0CB5C7A3-AD70-44F8-95DD-80C013EA4D2C/alert/text -- http://localhost:8100/ with parameters {
    wildcards =     (
        "session/0CB5C7A3-AD70-44F8-95DD-80C013EA4D2C/alert/text"
    );
}
```

## Environment

* Appium version (or git revision) that exhibits the issue: Appium v1.6.3 (REV 9cd5a21b473f0271b162bbe7f77f7d1468c0cfa1)
* Last Appium version that did not exhibit the issue (if applicable):
* Desktop OS/version used to run Appium: Mac OS Sierra 10.12.1
* Node.js version (unless using Appium.app|exe): v7.2.1
* Mobile platform/version under test: iOS 10.2
* Real device or emulator/simulator: iPhone 6 simulator
* Appium CLI or Appium.app|exe: Appium-Python-Client==0.24

## Details

If I have an alert up in the iOS simulator  (`XCUIElementTypeAlert`), and I run the following code:

```python
alert = self.driver.switch_to_alert()
alert.send_keys("something")
```

I will get the following error:

```
*** WebDriverException: Message: An unknown server-side error occurred while processing the command. Original error: Unhandled endpoint: /session/0CB5C7A3-AD70-44F8-95DD-80C013EA4D2C/alert/text -- http://localhost:8100/ with parameters {
    wildcards =     (
        "session/0CB5C7A3-AD70-44F8-95DD-80C013EA4D2C/alert/text"
    );
}
```

## Link to Appium logs

https://gist.github.com/SalmonMode/b947a01ca1f65272daddda99cb2a1773

## Code To Reproduce Issue [ Good To Have ]

**Comment by jlipps** [comment]
It looks like WDA doesn't support sendKeys on alerts. I believe it should be possible to find the text field in the alert using the normal find element route, however (i.e., not using `switch_to_alert`). Can you try that?

**Comment by mykola-mokhnach** [comment]
for me it just worked if I locate text input field on the alert by xpath and then use _setValue_ instead of _sendKeys_

**Comment by SalmonMode** [comment]
@jlipps If I try to select it based on its class and then call `send_keys` on that, it gives me this error:

```
   Traceback (most recent call last):
    test_compare_pharmacy_view.py line 706 in test_search
      self.change_med_search_near_location()
    onerx_base_tests.py line 1599 in change_med_search_near_location
      text_field.send_keys(location)
    /Users/cnejame/truveris/projects/arceus/venv/lib/python2.7/site-packages/selenium/webdriver/remote/webelement.py line 347 in send_keys
      self._execute(Command.SEND_KEYS_TO_ELEMENT, {'value': keys_to_typing(value)})
    /Users/cnejame/truveris/projects/arceus/venv/lib/python2.7/site-packages/selenium/webdriver/remote/webelement.py line 494 in _execute
      return self._parent.execute(command, params)
    /Users/cnejame/truveris/projects/arceus/venv/lib/python2.7/site-packages/selenium/webdriver/remote/webdriver.py line 236 in execute
      self.error_handler.check_response(response)
    /Users/cnejame/truveris/projects/arceus/venv/lib/python2.7/site-packages/appium/webdriver/errorhandler.py line 29 in check_response
      raise wde
   NoSuchElementException: Message: An element could not be located on the page using the given search parameters.
```

It is actually able to find the element, but i only get that error when I call `send_keys`

here's the log for that https://gist.github.com/SalmonMode/363523d7a17aa1a9cb64fd3088dba350

**Comment by SalmonMode** [comment]
@mykola-mokhnach I get the same thing using `set_value` as `send_keys` when selecting by class name.

**Comment by jlipps** [comment]
@SalmonMode weird, I'd reach out to the WebDriverAgent project then, as I believe this is supposed to work.

**Comment by mykola-mokhnach** [comment]
@SalmonMode Can you provide the code snippet where you try to find and set the value for the alert input?

**Comment by SalmonMode** [comment]
@mykola-mokhnach 

```python
text_field = self.driver.find_element_by_class_name("XCUIElementTypeTextField")
text_field.set_value("Edison, NJ")
```

**Comment by mykola-mokhnach** [comment]
@jlipps I believe the problem should be here:

appium-xcuitest-driver/lib/element.js

```javascript
commands.setValueImmediate = async function (value, el) {
  // WDA does not provide no way to set the value directly
  log.info('There is currently no way to bypass typing using XCUITest. Setting value through keyboard');
  await this.setValue(value, el);
};
```

WDA, actually, has setValue method. And since Appium itself still tries to use sendKeys instead WDA says "no way", because the keyboard is obstructed, which we, actually see in the log:

```
[debug] [JSONWP Proxy] Proxying [POST /element] to [POST http://localhost:8100/session/CEF01274-8C65-4EE6-A446-086A0C98270C/element] with body: {"using":"class name","value":"XCUIElementTypeKeyboard"}
[debug] [JSONWP Proxy] Got response with status 200: {"value":"Alert is obstructing view","sessionId":"CEF01274-8C65-4EE6-A446-086A0C98270C","status":26}
[debug] [BaseDriver] Waiting up to 0 ms for condition
```

I don't see this problem, because I'm smart ass and calling WDA's _/element/:uuid/value_ API directly.

**Comment by jlipps** [comment]
@mykola-mokhnach the ` /element/:uuid/value` endpoint is exactly what `await this.setValue(value, el)` is calling. So the problem must be that the keyboard is obstructed. @SalmonMode maybe try tapping in the text field once to bring up the keyboard? This is presumably what a user would have to do in this scenario.

**Comment by mykola-mokhnach** [comment]
@jlipps The further code in this file explicitly check for keyboard presence (check the very first condition). Appium should, probably, ignore an exception happening there and continue to setvalue.

```javascript
// make sure there is a keyboard if this is a text field
    if (await this.getAttribute('type', el) === 'XCUIElementTypeTextField') {
      await this.bringUpKeyboard(el);
    }

    if (value instanceof Array) {
      value = value.join('');
    }
    if (typeof value !== 'string') {
      value = value.toString();
    }

    if (!hasSpecialKeys(value)) {
      // nothing special, so just send it in
      await this.proxyCommand(`/element/${el}/value`, 'POST', {value: value.split('')});
      return;
    }
```

@SalmonMode you can try to workaround the issue by commenting this condition inside **commands.setValue** method of the corresponding file. Execute

```bash
gulp watch
``` 

inside /usr/local/lib/node_modules/appium/node_modules/appium-xcuitest-driver (node may ask you to install additional modules) to transpile your changes and restart your automated test.

**Comment by SalmonMode** [comment]
@jlipps the keyboard actually shows up when the alert pops up, but even when I click it before trying to type, the same thing happens (`Alert is obstructing view`). 

@mykola-mokhnach I noticed your fix just got merged, so I'll give that a shot.

**Comment by mykola-mokhnach** [comment]
Can you provide more details? which alerts? what is the diff between these alerts where it works and where it does not work? How the log looks like?

**Comment by squeemish** [comment]
All alerts are UIAlertControllerStyleAlert.

I have two links on a login page (request new activation mail and request new password) that open a UIAlertControllerStyleAlert with a single text field XCUIElementTypeTextField.

Both of those links give the `Alert is obstructing view` error.

Inside the application on a profile screen there is a Deactivate button that brings up a UIAlertControllerStyleAlert with a XCUIElementTypeTextField.  This alert takes the setValue command without issue.  The same for creating a group within the app.  

So far it's just the two alerts on the login page that exhibit the issue.  I will gather logs and attach later when I have them.

**Comment by imurchie** [comment]
Are there textfields behind the failing alert, and not behind the working one?

**Comment by squeemish** [comment]
yes.  there is one XCUIElementTypeTextField and one XCUIElementTypeSecureTextField behind the failing alert.

using the absolute XPath worked instead of the class locator.  Was attempting to grab the other text field.

Still shaking off the holiday I guess.  Thanks for the help...sorry for the false alarm

**Comment by squeemish** [comment]
as a side note, interesting that the locators I was using (find by class name) worked fine before this new version, but now fail.  Either way, i can use the XPath for these two alerts since the old locators are not behaving the same

**Comment by ABahadori156** [comment]
Could one of you guys paste your code for using absolute XPath to locate the text field inside the alert view and have it input text? I'd appreciate it - been stumped on this for a little as well

**Comment by squeemish** [comment]
@ABahadori156 

`private IOSElement getForgotPasswordPopupTextField() {
        return DriverUtils.findElementByXPath(driver, "//XCUIElementTypeApplication[1]/XCUIElementTypeWindow[1]/XCUIElementTypeOther[2]/XCUIElementTypeAlert[1]/XCUIElementTypeOther[1]/XCUIElementTypeOther[1]/XCUIElementTypeOther[2]/XCUIElementTypeOther[1]/XCUIElementTypeOther[1]/XCUIElementTypeOther[1]/XCUIElementTypeOther[1]/XCUIElementTypeCollectionView[1]/XCUIElementTypeCell[1]/XCUIElementTypeTextField[1]");
    }`

`public void enterEmailForPasswordReset(String resetPasswordEmail) {
        getForgotPasswordPopupTextField().setValue(resetPasswordEmail);
    }`

**Comment by ABahadori156** [comment]
Thanks so much!

Gonna give it a try now

**Comment by ABahadori156** [comment]
I am also getting the "Alert is obstructing view" error following with a 'No keyboard found. Clicking element to open it.'

For the XPath I am using:
private IOSElement alertTextFieldInput() { 

return (IOSElement) driver.findElementByXPath("//XCUIElementTypeApplication[1]/XCUIElementTypeOther[2]/XCUIElementTypeOther[2]/XCUIElementTypeAlert[1]/XCUIElementTypeOther[1]/XCUIElementTypeOther[1]/XCUIElementTypeOther[2]/XCUIElementTypeOther[1]/XCUIElementTypeOther[1]/XCUIElementTypeOther[1]/XCUIElementTypeOther[1]/XCUIElementTypeCollectionView[1]/XCUIElementTypeCell[1]/XCUIElementTypeTextField[1]");

}

then within my following method I do:

public void alertViews() {

alertTextFieldInput().setValue("Hello");

}

There are no other text fields in the view besides the one presented in the alert view. (I'm using the UICatalog app then I go to > Alert Views > Text Entry and it pops up an Alert View with a single text field)

The keyboard does NOT show up and the text field inside the alert is highlighted when popped up but it doesn't input in the "Hello" text.

Do I maybe need to add driver.switchTo().alert(); before my .setValue method?

Thanks.

**Comment by cosminstirbu** [comment]
Any updates on this issue. Is this fixed in the 1.6.4beta release?

For me it's reproducing with both sendKeys and setValue on 1.6.3

**Comment by vikramvi** [comment]
@triager please close; as per last comment issue is fixed
